### ISSUE TITLE
Multiple primary keys creates unhelpful generated code errors

### ISSUE THREAD

**Issue by DrewFenwick** [digitallyinduced/ihp#989]
IHP's UI does little to stop you from creating a table with two primary keys.

Using the UI I was able to make [this schema](https://gist.github.com/DrewFenwick/093a8e2cd7d69c45f2155f5ba49ef3e6),
which caused the following errors to show up in HLS

```hs

**Comment by CSchank** [comment]
Instead of disallowing multiple keys (which is the equivalent of composite keys), shouldn't there just be better support for them? Or would this mean significant changes to the code generation system for types / queries?

**Comment by DrewFenwick** [comment]
Ah, I must admit I'm not nearly as well versed at databases as I would like. I wasn't aware that multiple fields in a primary key meant a composite key.

If it's a commonly used feature of RDBMS then as a user I think it should be supported if possible.

**Comment by CSchank** [comment]
@DrewFenwick Yes, after a bit of research basically SQL can only have one primary key at a time (makes sense), but if you define multiple columns as primary keys, that's how you specify a composite primary key. After doing a bit of research, I think I was also mixing up my terminology a bit before: a composite key isn't necessarily a primary key, it can be simply a candidate key. But either way, we definitely want to be able to specify multiple primary keys in the UI.

But, it definitely isn't working properly! I think your issue is related to the issues I've been having and discussing on Gitter. Marc has created an issue #992 which I believe covers this issue. So hopefully both our problems will be fixed when #992 is fixed. :)

**Comment by mpscholten** [comment]
I've pushed a couple of fixed for composite PKs today. Still a bit more todo.

Also just tested the above schema. The issue is that you have a column `id` on the `idea_parents` table that is not a primary key. When having multiple primary keys in IHP, right now the table cannot have a column named `id` (when using a composite PK IHP emulates that the table has a `id` column, that causes the above error because your table actually also has a manual `id` column).

Will be fixed in the future. A workaround might be to not use composite PKs or not to call the non-primary-key column `id`.

**Comment by CSchank** [comment]
Ah, I see! Thank you. For now we're just ignoring them and using the two columns as our pretend "primary key".

**Comment by mpscholten** [comment]
Yes exactly. This is very hacky and some APIs of IHP will not support this at all. E.g I just rewrote the `deleteRecord` function to work with multiple PKs https://github.com/digitallyinduced/ihp/commit/e6046fde690b487c1bae0eb2603eddbda794317f

The new `primaryKeyCondition` in that commit will also help with making other APIs aware of composite PKs. Right now IHP also mostly assumes that there's a column called `id`. With the `primaryKeyCondition` we can also relax that constraint.

**Comment by CSchank** [comment]
... or would it be `|> filterWhere (#id, (activityId, 0))`? Probably not, but just wondering.

**Comment by mpscholten** [comment]
Yes, that should be supported then 👍  Likely `|> filterWhere (#id, (activityId, 0))`

**Comment by CSchank** [comment]
Okay, so now `id` becomes something of a conceptual "symbol" of the PK rather than necessarily a physical column name.

**Comment by mpscholten** [comment]
It is actually already right now https://github.com/digitallyinduced/ihp/blob/master/IHP/SchemaCompiler.hs#L173 When there's no id column, the SchemaCompiler calls `compileHasFieldId` to emulate that

To avoid naming conflicts we could also use `#primaryKey` instead of `#id` as the general identifier

**Comment by CSchank** [comment]
I see, well yes maybe another name would be better. Would `pk` work? It's maybe not as obvious what it means though. 

Also, I just ran my project with the latest dev server and I'm getting this: 

```
Mentor/FrontController.hs:29:9: error:
    • Couldn't match type ‘Id' "users"’ with ‘UUID’
        arising from a use of ‘initAuthentication’
    • In a stmt of a 'do' block: initAuthentication @User
      In the expression:
        do setLayout defaultLayout
           initAutoRefresh
           initAuthentication @User
           initUserAdminInfo
      In an equation for ‘initContext’:
          initContext
            = do setLayout defaultLayout
                 initAutoRefresh
                 initAuthentication @User
                 ....
   |
29 |         initAuthentication @User
```

Is this related? Weird thing is that this was working just now, when I submitted my pagination PR. Have you committed since?

**Comment by CSchank** [comment]
I pulled the very latest and now I'm getting spammed with: 

```
build/Generated/Types.hs:89:22: error:
    • No instance for (Default (Id' "users"))
        arising from a use of ‘def’
    • In the first argument of ‘User’, namely ‘def’
      In the expression:
        User def def def Nothing def def def def def def def
      In an equation for ‘newRecord’:
          newRecord = User def def def Nothing def def def def def def def
   |
89 |     newRecord = User def def def Nothing def def def def def def def
``` 

for many types. You probably already know this. Just posting it in case it's helpful.

**Comment by mpscholten** [comment]
Yes, likely related. I'll check it right now

**Comment by mpscholten** [comment]
Just had the same issue with another project. The problem is that I changed the SchemaCompiler and unless you call `make -B .envrc` it will still use the old schema compiler to build `build/Generated/Types.hs`. After you've run that, it should work

**Comment by CSchank** [comment]
It was weird. The first time I ran `make -B .envrc` it didn't work. What did seem to work was running `main` from `...DevServer.hs` in `ghci` instead of running `main` from `:l Main`. Now `:l Main` also works. 👍

**Comment by CSchank** [comment]
> I pulled the very latest and now I'm getting spammed with:
> 
> ```
> build/Generated/Types.hs:89:22: error:
>     • No instance for (Default (Id' "users"))
>         arising from a use of ‘def’
>     • In the first argument of ‘User’, namely ‘def’
>       In the expression:
>         User def def def Nothing def def def def def def def
>       In an equation for ‘newRecord’:
>           newRecord = User def def def Nothing def def def def def def def
>    |
> 89 |     newRecord = User def def def Nothing def def def def def def def
> ```
> 
> for many types. You probably already know this. Just posting it in case it's helpful.

This same thing just happened to me on our "production" server, and the only thing that worked was doing this:

> It was weird. The first time I ran `make -B .envrc` it didn't work. What did seem to work was running `main` from `:l ...DevServer.hs` in `ghci` instead of running `main` from `:l Main`. Now `:l Main` also works. 👍

Seems that just `make -B .envrc` wasn't enough.

**Comment by MonaMayrhofer** [comment]
Is this still a problem after #1925?
